src/rust/ide/lib/view/src/graph_editor/component/project_name.rs

```diff
@@ -0,0 +1,327 @@
+
+    ensogl::define_shape_system! {
+            (style:Style, selection:f32) {
+                let bg_color = color::Rgba::new(0.0,0.0,0.0,0.000_001);
+                Plane().fill(bg_color).into()
+            }
+        }
+}
+
+
+
+// =================
+// === FrpInputs ===
+// =================
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct FrpInputs {
+    /// Rename the project.
+    pub name : frp::Source<String>,
+    /// Reset the project name to the one before editing.
+    pub cancel_editing : frp::Source
+}
+
+impl FrpInputs {
+    /// Create new FrpInputs.
+    pub fn new(network:&frp::Network) -> Self {
+        frp::extend! {network
+            def cancel_editing = source();
+            def name           = source();
+        }
+        Self{cancel_editing,name}
+    }
+}
+
+
+
+// ==================
+// === FrpOutputs ===
+// ==================
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct FrpOutputs {
+    /// Emits the new project name when it's renamed.
+    pub name : frp::Source<String>
+}
+
+impl FrpOutputs {
+    /// Create new FrpOutputs.
+    pub fn new(network:&frp::Network) -> Self {
+        frp::extend! {network
+            def name = source();
+        }
+        Self{name}
+    }
+}
+
+
+
+// ===========
+// === Frp ===
+// ===========
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct Frp {
+    pub inputs  : FrpInputs,
+    pub outputs : FrpOutputs,
+    pub network : frp::Network,
+}
+
+impl Deref for Frp {
+    type Target = FrpInputs;
+    fn deref(&self) -> &Self::Target {
+        &self.inputs
+    }
+}
+
+impl Default for Frp {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Frp {
+    /// Create new Frp.
+    pub fn new() -> Self {
+        let network = frp::Network::new();
+        let inputs  = FrpInputs::new(&network);
+        let outputs = FrpOutputs::new(&network);
+        Self{network,inputs,outputs}
+    }
+}
+
+
+
+// ==================
+// === Animations ===
+// ==================
+
+/// ProjectName's animations handlers.
+#[derive(Debug,Clone,CloneRef)]
+pub struct Animations {
+    opacity  : Animation<f32>,
+    position : Animation<Vector3<f32>>
+}
+
+impl Animations {
+    /// Create new animations handlers.
+    pub fn new(network:&frp::Network) -> Self {
+        let opacity = Animation::new(&network);
+        let position = Animation::new(&network);
+        Self{opacity,position}
+    }
+}
+
+
+
+// ========================
+// === ProjectNameModel ===
+// ========================
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct ProjectNameModel {
+    logger         : Logger,
+    animations     : Animations,
+    display_object : display::object::Instance,
+    view           : component::ShapeView<background::Shape>,
+    text_field     : TextField,
+    project_name   : Rc<RefCell<String>>,
+    name_output    : frp::Source<String>,
+}
+
+impl ProjectNameModel {
```

Comment:
Soehow I feel your implementation is leaking memory. This is the reason I marked it now as "request changes". Please test what happens if you remove it from graph editor after some key is pressed. Will it be disposed?

Reply:
Yes, I just tested it and it's dropped. It's tested on the branch `wip/dg/project_renaming_view_drop_test`.

Reply:
Still, I believe it leaks memory. Look at `let project_name = self.clone_ref();` which you then pass to a closure managed by `self` - there is a loop.

Reply:
@wdanilo you seem to be right. But I have two notes about it:

1. I don't know how `CloneRef` aren't implemented. But when I run the test on the `wip/dg/project_renaming_view_drop_test` branch, the drop function is called. But disregard this note, the following one is more important.
2. This workaround for keeping the text field in a single line and checking if the change was committed with a newline is the one responsible for leaking memory. There is no easy fix for it without refactoring `TextField`. My suggestion is to leave a `FIXME` note to implement it correctly when the new `TextField` component arrives.

And one last thing: It would be cool to have two different components like `MultiLineText` and `SingleLineText`, with one being a specialization of the other. I see many places where `SingleLineText` could be useful.